v2/pkg/core/engine.go

```diff
@@ -0,0 +1,57 @@
+package core
+
+import (
+	"github.com/projectdiscovery/nuclei/v2/pkg/protocols"
+	"github.com/projectdiscovery/nuclei/v2/pkg/types"
+)
+
+// Engine is an engine for running Nuclei Templates/Workflows.
+//
+// The engine contains multiple thread pools which allow using different
+// concurrency values per protocol executed.
+//
+// The engine does most of the heavy lifting of execution, from clustering
+// templates to leading to the final execution by the workpool, it is
+// handled by the engine.
+type Engine struct {
+	workPool     *WorkPool
+	options      *types.Options
+	executerOpts protocols.ExecuterOptions
+}
+
+// InputProvider is an input provider interface for the nuclei execution
+// engine.
+//
+// An example InputProvider is provided in form of hmap input provider.
+type InputProvider interface {
+	// Count returns the number of items for input provider
+	Count() int64
+	// Scan calls a callback function till the input provider is exhausted
+	Scan(callback func(value string))
```

Comment:
The description suggests that the function is rather a consumer, and not a callback that is executed after the logic is finished

Reply:
I meant renaming the parameter name to `consumer` instead of `callback`... but anyway, you can leave it like this for now.